File: Device/SFBClockDevice.m
/*
 * Copyright (c) 2020 Stephen F. Booth <me@sbooth.org>
 * See https://github.com/sbooth/SFBAudioEngine/blob/master/LICENSE.txt for license information
 */

#import "SFBClockDevice.h"
#import "SFBAudioObject+Internal.h"

#import "SFBSystemAudioObject.h"
#import "SFBCStringForOSType.h"

@implementation SFBClockDevice

+ (NSArray *)clockDevices
{
	return [[SFBSystemAudioObject sharedInstance] audioObjectArrayForProperty:kAudioHardwarePropertyClockDeviceList inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (instancetype)initWithAudioObjectID:(AudioObjectID)objectID
{
	NSParameterAssert(SFBAudioObjectIsClassOrSubclassOf(objectID, kAudioClockDeviceClassID));
	return [super initWithAudioObjectID:objectID];
}

- (instancetype)initWithClockDeviceUID:(NSString *)clockDeviceUID
{
	NSParameterAssert(clockDeviceUID != nil);

	AudioObjectPropertyAddress propertyAddress = {
		.mSelector	= kAudioHardwarePropertyTranslateUIDToClockDevice,
		.mScope		= kAudioObjectPropertyScopeGlobal,
		.mElement	= kAudioObjectPropertyElementMaster
	};

	AudioObjectID objectID = kAudioObjectUnknown;
	UInt32 specifierSize = sizeof(objectID);
	OSStatus result = AudioObjectGetPropertyData(kAudioObjectSystemObject, &propertyAddress, sizeof(clockDeviceUID), &clockDeviceUID, &specifierSize, &objectID);
	if(result != kAudioHardwareNoError) {
		os_log_error(gSFBAudioObjectLog, "AudioObjectGetPropertyData (kAudioHardwarePropertyTranslateUIDToClockDevice) failed: %d '%{public}.4s'", result, SFBCStringForOSType(result));
		return nil;
	}

	if(objectID == kAudioObjectUnknown) {
		os_log_error(gSFBAudioObjectLog, "Unknown audio clock device UID: %{public}@", clockDeviceUID);
		return nil;
	}

	return [self initWithAudioObjectID:objectID];
}

- (NSString *)clockDeviceUID
{
	return [self stringForProperty:kAudioClockDevicePropertyDeviceUID inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (NSNumber *)transportType
{
	return [self unsignedIntForProperty:kAudioClockDevicePropertyTransportType inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (NSNumber *)domain
{
	return [self unsignedIntForProperty:kAudioClockDevicePropertyClockDomain inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (NSNumber *)isAlive
{
	return [self unsignedIntForProperty:kAudioClockDevicePropertyDeviceIsAlive inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (NSNumber *)isRunning
{
	return [self unsignedIntForProperty:kAudioClockDevicePropertyDeviceIsRunning inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (NSNumber *)latency
{
	return [self unsignedIntForProperty:kAudioClockDevicePropertyLatency inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (NSArray *)controls
{
	return [self audioObjectArrayForProperty:kAudioClockDevicePropertyControlList inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (NSNumber *)sampleRate
{
	return [self doubleForProperty:kAudioClockDevicePropertyNominalSampleRate inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (NSArray *)availableSampleRates
{
	return [self audioValueRangeArrayForProperty:kAudioClockDevicePropertyAvailableNominalSampleRates inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (NSString *)description
{
	return [NSString stringWithFormat:@"<%@ 0x%x, \"%@\">", self.className, _objectID, self.clockDeviceUID];
}

@end

File: Device/SFBClockDevice.swift
/*
 * Copyright (c) 2020 Stephen F. Booth <me@sbooth.org>
 * See https://github.com/sbooth/SFBAudioEngine/blob/master/LICENSE.txt for license information
 */

import Foundation

extension ClockDevice {
	/// Returns the clock device UID
	/// - note: This corresponds to `kAudioClockDevicePropertyDeviceUID`
	/// - throws: An error if the property could not be retrieved
	public func clockDeviceUID() throws -> String {
		return try getProperty(.clockDeviceUID)
	}

	/// Returns the transport type
	/// - note: This corresponds to `kAudioClockDevicePropertyTransportType`
	/// - throws: An error if the property could not be retrieved
	public func transportType() throws -> AudioDevice.TransportType {
		return AudioDevice.TransportType(rawValue: UInt32(try getProperty(.clockDeviceTransportType) as UInt))!
	}

	/// Returns the clock domain
	/// - note: This corresponds to `kAudioClockDevicePropertyClockDomain`
	/// - throws: An error if the property could not be retrieved
	public func domain() throws -> UInt {
		return try getProperty(.clockDeviceClockDomain)
	}

	/// Returns `true` if the clock device is alive
	/// - note: This corresponds to `kAudioClockDevicePropertyDeviceIsAlive`
	/// - throws: An error if the property could not be retrieved
	public func isAlive() throws -> Bool {
		return try getProperty(.clockDeviceIsAlive) != 0
	}

	/// Returns `true` if the clock device is running
	/// - note: This corresponds to `kAudioClockDevicePropertyDeviceIsRunning`
	/// - throws: An error if the property could not be retrieved
	public func isRunning() throws -> Bool {
		return try getProperty(.clockDeviceIsRunning) != 0
	}

	/// Returns the latency
	/// - note: This corresponds to `kAudioClockDevicePropertyLatency`
	/// - throws: An error if the property could not be retrieved
	public func latency() throws -> UInt {
		return try getProperty(.clockDeviceLatency)
	}

	/// Returns the audio controls
	/// - note: This corresponds to `kAudioClockDevicePropertyControlList`
	/// - throws: An error if the property could not be retrieved
	public func controls() throws -> [AudioControl] {
		return try getProperty(.clockDeviceControlList) as [AudioObject] as! [AudioControl]
	}

	/// Returns the device sample rate
	/// - note: This corresponds to `kAudioClockDevicePropertyNominalSampleRate`
	/// - throws: An error if the property could not be retrieved
	public func sampleRate() throws -> Double {
		return try getProperty(.clockDeviceNominalSampleRate)
	}

	/// Returns the available sample rates
	/// - note: This corresponds to `kAudioClockDevicePropertyAvailableNominalSampleRates`
	/// - throws: An error if the property could not be retrieved
	public func availableSampleRates() throws -> [AudioValueRange] {
		return try getProperty(.clockDeviceAvailableNominalSampleRates)
	}
}
